### PR TITLE
Modify formatting for CLI when calling --help

### PR DESCRIPTION
**Pull request by myNameIsPatrick**
## Description

This PR modifies the formatting for various help commands in the CLI to generally look nicer. There's also a few auxiliary items I have modified which I'll enumerate below. First, the CLI output:

#### Before:

```
 $ hop --help
usage: hop [-h] [--version] <command> ...

optional arguments:
  -h, --help  show this help message and exit
  --version   show program's version number and exit

Commands:
  <command>
    configure
              configuration utilities
    publish   parse and publish messages
    subscribe
              receive and parse messages
    version   list all the dependencies' versions
```

```
 $ hop configure --help
usage: hop configure [-h] <command> ...

Configuration utilities.

optional arguments:
  -h, --help  show this help message and exit

Commands:
  <command>
    locate    display configuration path
    setup     set up configuration
```

#### After:

```
 $ hop --help
usage: hop [-h] [--version] <command> ...

optional arguments:
  -h, --help          show this help message and exit
  --version           show program's version number and exit

commands:
    configure         configuration utilities
    publish           publish messages
    subscribe         subscribe to messages
    version           list all the dependencies' versions
```

```
 $ hop configure --help
usage: hop configure [-h] <command> ...

Configuration utilities.

optional arguments:
  -h, --help          show this help message and exit

commands:
    locate            display configuration path
    setup             set up configuration
```

The two main changes is this removal of the `<command>` line between the commands and the header and the spacing between commands and the description. This is handled by an internal function in `hop/__main__.py`, `_SubcommandHelpFormatter` which modifies the format of the help when passed into `formatter_class` in an argparse-based parser instance. Going a bit into the technical side, argparse seems to have a strange interaction between the metavar kwarg and subparsers which I'm essentially reverting so the spacing looks the same as if you didn't specify this.

The other auxiliary changes I have made are in `configure.py`. A few changes in the docstrings were made to match the rest of the codebase, and I added the `required` flag in the configure subparser so that you can't call `hop configure` by itself. This removes the need to check explicitly in the main function, and will warn the user accordingly of the proper usage.

## Checklist

* [x] All new functions and classes are documented and adhere to Google doc style (3.8.3-3.8.6 of [this](http://google.github.io/styleguide/pyguide.html#383-functions-and-methods) document)
* [x] Add/update sphinx documentation with any relevant changes.
* [x] Add/update pytest-style tests in `/tests`, ensuring sufficient code coverage.
* [x] `make test` runs without errors.
* [x] `make lint` doesn't give any warnings.
* [x] `make format` doesn't give any code formatting suggestions.
* [x] `make doc` runs without errors and generated docs render correctly.
* [x] Check that CI pipeline run on this PR passes all stages.
* [ ] Review signoff by at least one developer.

NOTE: If this PR relates to a release, open and reference an issue with the Release checklist template.

### CODE REVIEW

File: hop/__main__.py [hunk 1]
```diff
@@ -8,22 +8,51 @@
 from . import version
 
 
-def append_subparser(subparser, cmd, func):
+def append_subparser(subparser, cmd, func, formatter=False):
```
Comment (spenczar) [inline hop/__main__.py:11]: A boolean for `formatter` seems slightly odd. I would have expected it to be a formatter class, so you pass in `formatter=_SubcommandHelpFormatter`.

File: hop/__main__.py [hunk 2]
```diff
@@ -8,22 +8,51 @@
...
+        )
+    else:
+        parser = subparser.add_parser(cmd, help=help_, description=desc)
 
     parser.set_defaults(func=func)
+
     return parser
 
 
+class _SubcommandHelpFormatter(argparse.RawDescriptionHelpFormatter):
```
Comment (shereenElSayed) [inline hop/__main__.py:40]: What do you think if this class is moved to cli.py? I think it fits more there

File: hop/__main__.py (hunk 1 above)
Comment (shereenElSayed) [inline hop/__main__.py:11]: Maybe we can change formatter parameter to apply_formate?

**Comment by myNameIsPatrick** [comment]
@spenczar, @shereenElSayed, I see two different approaches on how to improve how `formatter` is handled within `append_subparser()`:

1. Pass in a formatter class instead of a boolean, via `formatter` or possibly `formatter_class` to match the kwarg being passed to the parser.
2. Keep `formatter` as a boolean but change the name to `apply_format`.

I would be slightly in favor of the first as it's a bit more flexible but otherwise don't have any strong opinions. Thoughts?

File: hop/__main__.py (hunk 2 above)
Comment (myNameIsPatrick) [reply hop/__main__.py:40]: I agree with that. I found a technical issue with doing that however, which is kinda annoying. The main issue is that adding it to the `cli` module adds in a circular dependency as `__main__.py` imports from all modules defining a `_main()` function while the other modules import from `cli.py` to grab commonly used CLI utilities.

Comment (shereenElSayed) [reply hop/__main__.py:40]: hmm I see, what do you think of a utility module?

**Comment by shereenElSayed** [comment]
@myNameIsPatrick I'd go with the first option

File: hop/__main__.py (hunk 1 above)
Comment (myNameIsPatrick) [reply hop/__main__.py]: That's a good idea, I added this in the last batch of commits.

Comment (myNameIsPatrick) [reply hop/__main__.py]: I modified this in the latest batch of commits.

File: hop/__main__.py (hunk 2 above)
Comment (myNameIsPatrick) [reply hop/__main__.py]: I like it. I added this class and the subparser function to `hop.utils.cli` in the latest batch of commits.

**Pull request merged**